### PR TITLE
Allow to drain node with pod that has more than one Pod Disruption Budget

### LINKED ISSUE #75957: Cannot drain node with pod with more than one Pod Disruption Budget
**What would you like to be added**:
When attempting to drain a node that has a pod scheduled on it with more than one PodDisruptionBudget the drain fails with the following message:
"Err: This pod has more than one PodDisruptionBudget, which the eviction subresource does not support."

It would be great if this was actually supported. 

**Why is this needed**:
It is not possible to upgrade clusters that have any pods with more than one PodDisruptionBudget since they fail on drain.

### PR DESCRIPTION
**Pull request by tedyu**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
We accommodate draining node with pod with more than one PDB.
In EvictionREST#Create, in the first round, pass dry-run being true to EvictionREST#checkAndDecrement and see if there is any non-nill error returned by any PDB.

If there is no non-nil error, we can go through PDBs again without dry-run (if deletionOptions.DryRun indicates so).
If there is non-nil error, just short circuit and return that error.

**Which issue(s) this PR fixes**:
Fixes #75957

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/registry/core/pod/storage/eviction.go
Comment (mortent) [inline pkg/registry/core/pod/storage/eviction.go]: Do we need to retry all of the PDBs? I think we should do this inside the loop.

File: pkg/registry/core/pod/storage/eviction.go
Comment (mortent) [inline pkg/registry/core/pod/storage/eviction.go]: Can we make it clearer here that we are doing an initial run with dry-run=true followed by a run with dry-run=false?

File: pkg/registry/core/pod/storage/eviction.go
Comment (mortent) [inline pkg/registry/core/pod/storage/eviction.go]: We need to remove the check on the number of pdbs on line 149 for this to work.

**Review by mortent — CHANGES_REQUESTED** [review]
We should add tests for this

File: pkg/registry/core/pod/storage/eviction.go
Comment (tedyu) [reply pkg/registry/core/pod/storage/eviction.go]: Used bool variable for iteration.

**Comment by tedyu** [comment]
I am preparing new test for this.

Meanwhile, review from wider audience would help gauge the direction this PR takes.

**Comment by tedyu** [comment]
@mortent 
I have added a new test case.

File: pkg/registry/core/pod/storage/eviction.go
Comment (tedyu) [reply pkg/registry/core/pod/storage/eviction.go]: Any conflict error from participating PDBs is retried.

Comment (mortent) [reply pkg/registry/core/pod/storage/eviction.go]: So my concern here is that the call to `checkAndDecrement` is not idempotent. So if we have 2 pdbs and we get a conflict while updating the last one, we will end up running `checkAndDecrement` twice on the first one. This can lead to situations where a disruption is rejected when it should be allowed.

Would it be possible for us to check in `checkAndDecrement` whether the given pod is already in the `DisruptedPods` map and therefore has already been counted in `PodDisruptionsAllowed`? In this case, multiple calls to `checkAndDecrement` would be less of a concern.

Comment (tedyu) [reply pkg/registry/core/pod/storage/eviction.go]: Thanks, @mortent for the suggestion.
I have adopted in the latest push.

File: pkg/registry/core/pod/storage/eviction.go
Comment (mortent) [inline pkg/registry/core/pod/storage/eviction.go]: I don't think we can set refresh to true here. If the first PDB is successfully decremented, we might still need to refresh the remaining PDBs. If we have two PDBs and we get a conflict on the last one, we will retry both of them, set the refresh variable to true after the first one and then retry the second PDB with the old pdb, which will fail with conflict again.
I think it might be easier if we retry on a per-pdb basis instead of retrying all pdbs if we get a conflict on one of them. And once we have completed the first dry-run and start updating the PDBs, it shouldn't be necessary to do any more dry-runs.

File: pkg/registry/core/pod/storage/eviction.go
Comment (mortent) [inline pkg/registry/core/pod/storage/eviction.go]: I think we still want to update the timestamp in the DisruptedPods map even if the pod has already been counted in PodDisruptionsAllowed. Otherwise there is a chance the entry in the map might time out too early.

**Comment by tedyu** [comment]
@mortent @liggitt 
Please take another look.

File: pkg/registry/core/pod/storage/eviction.go
Comment (mortent) [inline pkg/registry/core/pod/storage/eviction.go]: Would this be better if we moved the retry logic inside the loop over PDBs? This means we will retry each PDB until we succeed instead of redoing all PDBs every time we get a conflict on either of them. It seems like the current approach could lead to unnecessary decrements on PDBs where there are no conflicts. 

Also, seems like the fix in checkAndDecrement to avoid decrementing the allowed disruptions multiple times for a single pod is no longer in this PR. Any reason why this is no longer here?

Comment (tedyu) [reply pkg/registry/core/pod/storage/eviction.go]: I have put back the fix in checkAndDecrement to avoid decrementing the allowed disruptions multiple times (which would address the unnecessary decrements on PDBs where there are no conflicts).

**Comment by tedyu** [comment]
@mortent @liggitt
Please review again.

File: pkg/registry/core/pod/storage/eviction.go
Comment (mortent) [inline pkg/registry/core/pod/storage/eviction.go]: I think we want to do this check only for deciding whether PodDisruptionsAllowed should be decremented. The update of the timestamp for the pod that happens on line 244 should probably happen anyway so pods are no prematurely removed from the DisruptedPods map.

File: pkg/registry/core/pod/storage/eviction.go
Comment (mortent) [reply pkg/registry/core/pod/storage/eviction.go]: Any reason why you prefer the retry outside the loop over the PDBs? With the fix in checkAndDecrement I don't think it should have any impact on whether an eviction is allowed, but I think it could lead to unnecessary fetch and update for PDBs that have already been updated.

Comment (tedyu) [reply pkg/registry/core/pod/storage/eviction.go]: 'matching pdbs with one disruption allowed and one disruption disallowed' case wouldn't pass when the retry logic is inside the loop

Comment (mortent) [reply pkg/registry/core/pod/storage/eviction.go]: Why? Can you walk through the scenario that would fail?

Comment (tedyu) [reply pkg/registry/core/pod/storage/eviction.go]: @mortent 
I have updated the PR and moved the retry inside.

**Comment by tedyu** [comment]
@mortent @liggitt 
Previous review comments have been addressed.
Please take another look.

File: pkg/registry/core/pod/storage/eviction.go
Comment (mortent) [inline pkg/registry/core/pod/storage/eviction.go]: The refresh is on a per-pdb basis, so I think this should be set inside the loop over the pdbs.

File: pkg/registry/core/pod/storage/eviction.go
Comment (mortent) [inline pkg/registry/core/pod/storage/eviction.go]: In which scenarios do we want to keep going if we get an error here? Most errors that could be returned from the checkAndDecrement function means that eviction of the pod is blocked by a pdb. In this case, it doesn't seem like we should keep trying to update the remaining pdbs.

**Comment by tedyu** [comment]
@mortent 
Your comments have been addressed.

**Comment by tedyu** [comment]
@mortent @liggitt
Please take another look.

thanks

File: pkg/registry/core/pod/storage/eviction.go
Comment (mortent) [inline pkg/registry/core/pod/storage/eviction.go]: We only want to decrement when we are seeing a pod that isn't already in the `DisruptedPods` map right? This only decrements for pods that are already in the map. Maybe add a test for this.

Comment (tedyu) [reply pkg/registry/core/pod/storage/eviction.go]: @mortent 
Please take another look.

Comment (tedyu) [reply pkg/registry/core/pod/storage/eviction.go]: The existing test uses REST API to simulate various scenarios.

Still need to figure out how to make assertion w.r.t. the decrement.

**Review by mortent — APPROVED** [review]
/lgtm

**Comment by tedyu** [comment]
@liggitt 
Please take a look.

**Comment by liggitt** [comment]
This is not the approach I expected. Doing a dry-run across all PDBs followed by a normal update across all PDBs:
* it doubles the number of write requests
* it is not well behaved in the face of concurrent eviction requests with multiple PDBs

I expected a work-sharing approach closer to the resourcequota admission plugin:
* https://github.com/kubernetes/kubernetes/blob/release-1.17/plugin/pkg/admission/resourcequota/controller.go#L587-L652
* https://github.com/kubernetes/kubernetes/blob/release-1.17/plugin/pkg/admission/resourcequota/admission.go#L134-L145

Such an approach would have the following characteristics:
* no increase in PDB writes
* concurrent eviction requests for the same namespace could be handled together, resulting in one write per PDB

**Comment by tedyu** [comment]
@liggitt 
```
it doubles the number of write requests
```
If you meant that the PDB would be updated twice, that shouldn't be the case.
The following is only executed after the dry-run check:
```
	if _, err := r.podDisruptionBudgetClient.PodDisruptionBudgets(namespace).UpdateStatus(&pdb); err != nil {
```
Otherwise, please clarify the scenario where the write is doubled.
Note: there may be more than two PDBs under consideration here.

**Comment by tedyu** [comment]
The current solution is loosely based on two-phase commit protocol.

* the dry-run corresponds to voting phase where each pdb decides either committing or aborting the procedure.
* the normal update corresponds to the commit-request phase only if all pdb's permit the action

**Comment by mml** [comment]
FYI I added some more food for thought to the underlying issue.  I think it could use a bit more discussion before merging a fix.  Thanks and sorry for the late comments.

**Comment by tedyu** [comment]
@mml 
Thanks for chiming in.
In the current PR, there is the following guard before DisruptionsAllowed is decremented.
{code}
+       if _, ok := pdb.Status.DisruptedPods[podName]; !ok {
+               pdb.Status.DisruptionsAllowed--
+       }
{code}
I am thinking of also adding a sort step (by pdb.Namespace, then by pdb.Name) prior to the nested loop in EvictionREST#Create so that the behavior is more predictable.

**Comment by tedyu** [comment]
Added a helper func:
```
func (r *EvictionREST) getDisruptionsAllowed(namespace string, pdbName string) (int32, error) {
```
using which assertion is made on DisruptionsAllowed not becoming negative.

**Comment by mortent** [comment]
So I think there are some issues with concurrent evictions we should think about here. Seems to me there are two situations:

**Concurrent evictions of the same pod**: I think @tedyu is right that the guard before decrementing `AllowedDisruptions` and sorting to make sure PDBs are processed in the same order should help here. But I think there is still possible that with two concurrent evictions, they will interact in a way that cause a disruption to be rejected when it actually should be allowed. I can see this happening if one call to the Eviction API updates the first PDB and then the second call updates the second PDB.  We might be able to fix this by checking earlier in the `checkAndDecrement` function whether the pod is already listed in the `DisruptedPods` map. Anyway, it seems like two concurrent evictions when successful will lead to two calls to terminate the pod as well as several additional reads due to retries.
Using a worker similar to the example in the ResourceQuota admission plugin as suggested by @liggitt seems like a good approach here. Then the second eviction request will just wait for the first one to finish.

**Concurrent evictions of different pods with one or more shared PDBs**: If two different nodes are being drained at the same time, it is possible that each of them will terminate the last pod of different PDBs, which would block termination of other pods belonging to the same PDBs. This is in some ways the desired behavior since we will avoid draining nodes in a way that violates the PDBs. I think this situation is possible today even without allowing multiple PDBs per pod.


Following up on the comment from @mml, do we have any compelling use cases for when it is useful to have multiple PDBs per pod? I agree that it can make it very hard to understand what is going on.

**Comment by tedyu** [comment]
My understanding for using a worker similar to the example in the ResourceQuota admission plugin is for the request submissions.
Inside EvictionREST#Create, the two pass approach is able to handle the multiple PDBs in the current formation.

**Comment by annajung** [comment]
Hello @mortent @liggitt @lavalamp  👋 I'm one of the Bug Triage shadows for the 1.18 release. Since this issue is tagged for 1.18, I'd like to check its status. Is this issue intended for milestone 1.18?

Friendly reminder: Code freeze is scheduled for March 5th.

**Comment by lavalamp** [comment]
I'd assume this is not making it at this point, the comments don't seem to have converged (and I don't have time to review it right now, sorry).

**Comment by annajung** [comment]
@lavalamp Thanks for the response! I'll defer the milestone to 1.19.
/milestone v1.19

**Comment by tedyu** [comment]
In favor of #90253

**Pull request closed without merging**
